src/main/java/edu/harvard/iq/dataverse/authorization/providers/oauth2/OAuth2LoginBackingBean.java

```diff
@@ -56,20 +62,78 @@
+                  .scope(idp.getScope())
+                  .build();
     }
-
+    
+    /**
+     * View action for callback.xhtml, the browser redirect target for the OAuth2 provider.
+     * @throws IOException
+     */
     public void exchangeCodeForToken() throws IOException {
-        HttpServletRequest req = (HttpServletRequest) FacesContext.getCurrentInstance().getExternalContext().getRequest();
+        HttpServletRequest req = (HttpServletRequest) facesCtx.getExternalContext().getRequest();
+        
+        try {
+            Optional<AbstractOAuth2AuthenticationProvider> oIdp = parseStateFromRequest(req);
+            Optional<String> code = parseCodeFromRequest(req);
 
-        final String code = req.getParameter("code");
+            if (oIdp.isPresent() && code.isPresent()) {
+                AbstractOAuth2AuthenticationProvider idp = oIdp.get();
+                
+                OAuth20Service svc = idp.getService(systemConfig.getOAuth2CallbackUrl());
+                oauthUser = idp.getUserRecord(code.get(), svc);
+                
+                UserRecordIdentifier idtf = oauthUser.getUserRecordIdentifier();
+                AuthenticatedUser dvUser = authenticationSvc.lookupUser(idtf);
+    
+                if (dvUser == null) {
+                    // need to create the user
+                    newAccountPage.setNewUser(oauthUser);
+                    FacesContext.getCurrentInstance().getExternalContext().redirect("/oauth2/firstLogin.xhtml");
+        
+                } else {
+                    // login the user and redirect to HOME of intended page (if any).
+                    session.setUser(dvUser);
+                    session.configureSessionTimeout();
+                    final OAuth2TokenData tokenData = oauthUser.getTokenData();
+                    tokenData.setUser(dvUser);
+                    tokenData.setOauthProviderId(idp.getId());
+                    oauth2Tokens.store(tokenData);
+                    String destination = redirectPage.orElse("/");
+                    HttpServletResponse response = (HttpServletResponse) FacesContext.getCurrentInstance().getExternalContext().getResponse();
+                    String prettyUrl = response.encodeRedirectURL(destination);
+                    FacesContext.getCurrentInstance().getExternalContext().redirect(prettyUrl);
+                }
+            }
+        } catch (OAuth2Exception ex) {
+            error = ex;
+            logger.log(Level.INFO, "OAuth2Exception caught. HTTP return code: {0}. Message: {1}. Message body: {2}", new Object[]{error.getHttpReturnCode(), error.getLocalizedMessage(), error.getMessageBody()});
+            Logger.getLogger(OAuth2LoginBackingBean.class.getName()).log(Level.SEVERE, null, ex);
+        } catch (InterruptedException | ExecutionException ex) {
+            error = new OAuth2Exception(-1, "Please see server logs for more details", "Could not login due to threading exceptions.");
```

Comment:
Same question as above. Does this English ever show up in the web interface for the end user?

Reply:
Most certainly not. Won't change...